acceptance: replica allocator tests

cuongdo opened the pull request:
Adds these tests:
- Up-replication of 10 GiB from 1 to 3 nodes
- Rebalancing of 10 GiB unique data from 3 to 5 nodes
- Up-replication of 108 GiB from 1 to 3 nodes

These are all necessarily remote tests that use `terrafarm`. Current
limitations include:
1. SSH operations are all done in serial, including downloading of
   archived store data and copying of custom `cockroach` binaries.
2. Lacks ability to interrupt test at precise points in time, which is
   helpful for debugging and extending tests.
3. Doesn't provide conveniences for manually intervening in a test
   cluster (e.g. pushing new binaries mid-test, stopping/starting
   cluster, SSH'ing into individual machines, opening admin UI).
4. No automated way to archive stores to Google Cloud Storage.
5. Control-C behavior is a little weird. Pressing Control-C seems to
   abort just the current part of the test, not the whole test.

These will be fixed in later PRs.

cc @tschottdorf (Terrafarm)

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7315)

<!-- Reviewable:end -->

Commit:
acceptance: replica allocator tests

Adds these tests:
- Up-replication of 10 GiB from 1 to 3 nodes
- Rebalancing of 10 GiB unique data from 3 to 5 nodes
- Up-replication of 108 GiB from 1 to 3 nodes

These are all necessarily remote tests that use `terrafarm`. Current
limitations include:

1. SSH operations are all done in serial, including downloading of
   archived store data and copying of custom `cockroach` binaries.
2. Lacks ability to interrupt test at precise points in time, which is
   helpful for debugging and extending tests.
3. Doesn't provide conveniences for manually intervening in a test
   cluster (e.g. pushing new binaries mid-test, stopping/starting
   cluster, SSH'ing into individual machines, opening admin UI).
4. No automated way to archive stores to Google Cloud Storage.
5. Control-C behavior is a little weird. Pressing Control-C seems to
   abort just the current part of the test, not the whole test.

These will be fixed in later PRs.

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (I'm going to go ahead and try to run it locally next).

---

Reviewed 13 of 13 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[acceptance/allocator_test.go, line 207 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiKxTccCQvPs2Lw2SH:-KKiKxTccCQvPs2Lw2SI:416872728) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L207)):*
> ```Go
> // been stable for at least `stableInterval`.
> func (at *allocatorTest) checkAllocatorStable(db *gosql.DB) (bool, error) {
> 	q := `SELECT NOW() - MAX(timestamp) FROM rangelog`
> ```

Should this be more specific (i.e. only check range log events which relate to what you care about - probably splits and rebalances)?

---

*[acceptance/allocator_test.go, line 209 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiKoyANrZ1Jxt0ZNpp:-KKiKoyANrZ1Jxt0ZNpq:2126619823) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L209)):*
> ```Go
> 	q := `SELECT NOW() - MAX(timestamp) FROM rangelog`
> 	var elapsedStr string
> 	if err := db.QueryRow(q).Scan(&elapsedStr); err != nil {
> ```

You can't scan into a time.Duration?

---

*[acceptance/allocator_test.go, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiL9dY5yWH_rf6xibw:-KKiL9dY5yWH_rf6xibx:-1395735780) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L222)):*
> ```Go
> 	var status string
> 	stable := elapsedSinceLastRangeEvent >= stableInterval
> 	if stable {
> ```

nit: you could inline `stable`.

---

*[acceptance/allocator_test.go, line 232 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiLQfCsyJ-VokVf7yR:-KKiLQfCsyJ-VokVf7yS:1327158486) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L232)):*
> ```Go
> 
> // WaitForRebalance waits until there's been no recent range adds, removes, and
> // splits. We wait until the cluster is balanced or until `maxWait` elapses,
> ```

Afaict, this is the only mention of `maxWait` (probably want `stableInterval` - it should be exported if you mention it here and it's constant, since this method is exported).

---

*[acceptance/allocator_test.go, line 246 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiLxwULjjy_s03iotF:-KKiLxwVL6XZu4x46hV8:753238455) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L246)):*
> ```Go
> 	}()
> 
> 	c := make(chan os.Signal, 1)
> ```

I think we already have a channel that does this: `stopper`. It's not a great pattern but in this case, you could just add `<-stopper` to your select to the same effect.

---

*[acceptance/util_test.go, line 64 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiMPoXENJ_rRyJq2U5:-KKiMPoXENJ_rRyJq2U6:1874074915) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/util_test.go#L64)):*
> ```Go
> 
> // prefixRE is extracted from a Terraform error message regarding invalid
> // resource names.
> ```

Could you add an example of the message here?

---

*[acceptance/allocator_terraform/load.tf, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiMovRAFmPuQ6rkC9F:-KKiMovS90UxsiRtrz1W:645242527) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_terraform/load.tf#L1)):*
> ```HCL
> # Load generators that can be used to fill the CockroachDB cluster with data.
> ```

Is this mostly c&p from our other terraform stuff?

---

*[acceptance/allocator_terraform/nodectl, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiN1T-ITtg11ADRS_v:-KKiN1T-ITtg11ADRS_w:1318179808) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_terraform/nodectl#L1)):*
> ```
> #! /bin/bash
> ```

`#!/usr/bin/env bash`

---

*[acceptance/allocator_terraform/nodectl, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiN2SxnY3l3zHhq9BY:-KKiN2SxnY3l3zHhq9BZ:-787876604) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_terraform/nodectl#L3)):*
> ```
> #! /bin/bash
> 
> set -eu -o pipefail
> ```

nit: `set -euo pipefail`.

---

*[acceptance/allocator_terraform/nodectl, line 24 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiN9wK9QVzzyHQorF_:-KKiN9wK9QVzzyHQorFa:585360075) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_terraform/nodectl#L24)):*
> ```
> # CockroachDB nodes are named blah-cockroach-[0-9]*, so extract the final
> # numeric part.
> gcs_url=$2
> ```

nit: `var="${something}"` throughout

---

*[acceptance/allocator_terraform/output.tf, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiNIj6BlysfOXKmF7J:-KKiNIj6BlysfOXKmF7K:-1386969520) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_terraform/output.tf#L1)):*
> ```HCL
> #output "cockroach_ips" {
> ```

what's with the commented out parts (here and above)?

---

*[acceptance/allocator_terraform/variables.tf, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiNNIbneKNj8P87I4d:-KKiNNIbneKNj8P87I4e:1137242453) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_terraform/variables.tf#L1)):*
> ```HCL
> # Number of CockroachDB instances. This is overridden by clusterctl.
> ```

What's `clusterctl`?

---

*[acceptance/terrafarm/farmer.go, line 292 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiNvnrM1kLNuQ25njO:-KKiNvnsXOH4x1wltRY5:-1217025017) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/terrafarm/farmer.go#L292)):*
> ```Go
> }
> 
> // StartCockroach starts CockroachDB on all nodes.
> ```

`RestartAll`, like the next comment below (or just use

```go
for i := 0; i < f.NumNodes(); i++ {
  if err := f.Restart(i); err != nil {
     // xx
  }
}
```
)

---

*[acceptance/terrafarm/farmer.go, line 302 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiNYe4_R1naaTnisCo:-KKiNYe4_R1naaTnisCp:258593451) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/terrafarm/farmer.go#L302)):*
> ```Go
> }
> 
> // StopCockroach stops CockroachDB on all nodes.
> ```

Why not use `Kill()`? You could change the interface so that it takes a boolean `graceful`. If you want to save the boilerplate for iterating over all nodes, you could have a helper `Killall(graceful bool)`.

---

*[acceptance/terrafarm/farmer.go, line 312 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiOGAFn7vozQUCpnMc:-KKiOGAFn7vozQUCpnMd:-202293396) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/terrafarm/farmer.go#L312)):*
> ```Go
> }
> 
> // SSHCockroachNodes executes the given command on all CockroachDB nodes.
> ```

Ditto, this is very close to `ExecRoot` so `ExecRootAll` or remove.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-:-KKiOc_l4b4eIQmIjFcp:-1010720725)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with some open comments (and all of Tobi's)

I also didn't look at the terraform files.

---

Reviewed 3 of 13 files at r1.
Review status: all files reviewed at latest revision, 22 unresolved discussions, all commit checks successful.

---

*[acceptance/allocator_test.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiiN7swsAz3KW9EET0:-KKiiN7swsAz3KW9EET1:-690541000) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L53)):*
> ```Go
> 	"time"
> 
> 	"github.com/montanaflynn/stats"
> ```

Do you need this?  Don't we already compute our own?

---

*[acceptance/allocator_test.go, line 85 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiRqf0JG3qMKr6tewK:-KKiRqf1xaa_QRl2XWnL:1878223126) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L85)):*
> ```Go
> 	// for CockroachDB nodes. Leaving this as 0 accepts the default in the
> 	// Terraform configs.
> 	CockroachDiskSizeGB int
> ```

We tend to use bytes instead of GB.  And int64 to hold the values.

---

*[acceptance/allocator_test.go, line 101 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiS5mwZHpV0TUZVDq9:-KKiS5mwZHpV0TUZVDqA:7014757) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L101)):*
> ```Go
> 	}
> 	if at.CockroachDiskSizeGB != 0 {
> 		at.f.AddVars["cockroach_disk_size"] = strconv.Itoa(at.CockroachDiskSizeGB)
> ```

If If you make this bytes, you can use humanizeutil's ParseBytes and you'll be able to input a large number of formats. 

---

*[acceptance/allocator_test.go, line 112 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiSkm5hjrJrZdY837n:-KKiSkm5hjrJrZdY837o:750847356) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L112)):*
> ```Go
> 	log.Info("initial cluster is up")
> 
> 	log.Info("stopping CockroachDB")
> ```

perhaps just elaborate a bit on why you're stopping the cluster in the log

---

*[acceptance/allocator_test.go, line 147 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiT_kVYQ86lCCvh1Dx:-KKiT_kVYQ86lCCvh1Dy:259189464) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L147)):*
> ```Go
> // printStats prints the time it took for rebalancing to finish and the final
> // standard deviation of replica counts across stores.
> func (at *allocatorTest) printRebalanceStats(db *gosql.DB, host string, adminPort int) error {
> ```

Not something to work on now, but I can see a need to print a collection of different stats from the cluster (probably via command line arguments),  and ideally graph them over time as rebalancing proceeds.

---

*[acceptance/allocator_test.go, line 151 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiT4OUOdcTfFCqbanS:-KKiT4OUOdcTfFCqbanT:-837916475) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L151)):*
> ```Go
> 
> 	// Output time it took to rebalance.
> 	{
> ```

Is there a need for these scopes?  Or it is just to separate the 3 outputs?  

---

*[acceptance/allocator_test.go, line 246 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiLxwULjjy_s03iotF:-KKiWLTzZktBcJaQ5Tpm:-510466975) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L246)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I think we already have a channel that does this: `stopper`. It's not a great pattern but in this case, you could just add `<-stopper` to your select to the same effect.
</details>
:+1: 

---

*[acceptance/allocator_test.go, line 312 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiWbaVbPNSlEFXnkBu:-KKiWbaVbPNSlEFXnkBv:968937252) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L312)):*
> ```Go
> 		Prefix:              "uprep-1to3-med",
> 		RebalanceTimeout:    4 * time.Hour,
> 		CockroachDiskSizeGB: 250, // GiB
> ```

So is it in GiB or GB?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-:-KKijOTgye9TPxr2RtgB:-1311376245)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 7 of 13 files reviewed at latest revision, 22 unresolved discussions, some commit checks pending.

---

*[acceptance/allocator_test.go, line 53 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiiN7swsAz3KW9EET0:-KKj3HXFTiJctXFHW9iD:-382804123) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L53)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Do you need this?  Don't we already compute our own?
</details>
Where? Specifically, I need standard deviation. We have something in `StorePool` I believe, but that's a running standard deviation, so it isn't as accurate.

---

*[acceptance/allocator_test.go, line 85 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiRqf0JG3qMKr6tewK:-KKj3UiYTSW5Y2R3lwua:163430342) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L85)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> We tend to use bytes instead of GB.  And int64 to hold the values.
</details>
This is passed directly to Terraform, which expects GB. If I changed this to bytes, there'd be a loss of fidelity passing it to Terraform anyway. I've clarified this in the comment.

---

*[acceptance/allocator_test.go, line 101 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiS5mwZHpV0TUZVDq9:-KKj3f2GM1jzDemKFUyK:-1612024844) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L101)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> If If you make this bytes, you can use humanizeutil's ParseBytes and you'll be able to input a large number of formats. 
</details>
This won't work for the reasons mentioned above.

---

*[acceptance/allocator_test.go, line 112 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiSkm5hjrJrZdY837n:-KKj3hGGdW8BZWwpHhca:-220745012) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L112)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> perhaps just elaborate a bit on why you're stopping the cluster in the log
</details>
This part has been removed.

---

*[acceptance/allocator_test.go, line 151 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiT4OUOdcTfFCqbanS:-KKj3ldAHDiLbpPwAxvQ:126224165) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L151)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Is there a need for these scopes?  Or it is just to separate the 3 outputs?  
</details>
It's so that I don't have to come up with different names for common variables like `q`. It keeps things a bit more succinct and prevents unintentional sharing of state between the 3 outputs.

---

*[acceptance/allocator_test.go, line 207 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiKxTccCQvPs2Lw2SH:-KKiQiLrBQgkIVu2r_lU:1460369365) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L207)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Should this be more specific (i.e. only check range log events which relate to what you care about - probably splits and rebalances)?
</details>
Up-replications (1->3 nodes) are replica adds. Rebalances are adds followed by removes. I *think* splits are relevant, because they tend to trigger adds, but maybe I don't need to concern myself with them, since a lack of replica adds following a split will show up in a higher standard deviation of range counts.

---

*[acceptance/allocator_test.go, line 209 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiKoyANrZ1Jxt0ZNpp:-KKj43JKMwWqKHdyMf7B:-2102245820) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L209)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> You can't scan into a time.Duration?
</details>
I could've sworn this gave me an error earlier, but it works now. Done.

---

*[acceptance/allocator_test.go, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiL9dY5yWH_rf6xibw:-KKj4H07CWnb3OhvLuY4:-1141207135) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L222)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> nit: you could inline `stable`.
</details>
I'm returning `stable` below though

---

*[acceptance/allocator_test.go, line 232 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiLQfCsyJ-VokVf7yR:-KKiqYBhAm8UsP7urLSA:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L232)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Afaict, this is the only mention of `maxWait` (probably want `stableInterval` - it should be exported if you mention it here and it's constant, since this method is exported).
</details>
Done.

---

*[acceptance/allocator_test.go, line 246 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiLxwULjjy_s03iotF:-KKiqigZtL7R5IjViTcA:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L246)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> :+1: 
</details>
Done.

---

*[acceptance/allocator_test.go, line 312 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiWbaVbPNSlEFXnkBu:-KKiqm5KIEnZf_6hMrZZ:-7987711) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L312)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> So is it in GiB or GB?
</details>
GB, because that's what Terraform uses. Fixed.

---

*[acceptance/util_test.go, line 64 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiMPoXENJ_rRyJq2U5:-KKjG64_wwVF2wkh1ttP:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/util_test.go#L64)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Could you add an example of the message here?
</details>
Done.

---

*[acceptance/allocator_terraform/load.tf, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiMovRAFmPuQ6rkC9F:-KKj4OXJRA84ud9oRuIQ:-663446168) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_terraform/load.tf#L1)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Is this mostly c&p from our other terraform stuff?
</details>
Yes, except with more configurability and some extra packages not needed for other tests (e.g. google-cloud-sdk).

---

*[acceptance/allocator_terraform/nodectl, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiN1T-ITtg11ADRS_v:-KKixTHyKLmOHo_oJgcZ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_terraform/nodectl#L1)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `#!/usr/bin/env bash`
</details>
Done.

---

*[acceptance/allocator_terraform/nodectl, line 3 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiN2SxnY3l3zHhq9BY:-KKixWUigKoF6mtzwXIo:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_terraform/nodectl#L3)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> nit: `set -euo pipefail`.
</details>
Done.

---

*[acceptance/allocator_terraform/nodectl, line 24 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiN9wK9QVzzyHQorF_:-KKixmHF9-lDaNK_tqF4:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_terraform/nodectl#L24)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> nit: `var="${something}"` throughout
</details>
Done.

---

*[acceptance/allocator_terraform/output.tf, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiNIj6BlysfOXKmF7J:-KKixtsHXbLNYCaKJrVd:58936839) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_terraform/output.tf#L1)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> what's with the commented out parts (here and above)?
</details>
cruft from a previous revision. deleted.

---

*[acceptance/allocator_terraform/variables.tf, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiNNIbneKNj8P87I4d:-KKiy0h8IQLnh6jktmGE:499967786) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_terraform/variables.tf#L1)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What's `clusterctl`?
</details>
something that doesn't exist any more. changed.

---

*[acceptance/terrafarm/farmer.go, line 292 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiNvnrM1kLNuQ25njO:-KKj2BsuCItM7UIIGlBz:460095359) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/terrafarm/farmer.go#L292)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `RestartAll`, like the next comment below (or just use
> 
> ```go
> for i := 0; i < f.NumNodes(); i++ {
>   if err := f.Restart(i); err != nil {
>      // xx
>   }
> }
> ```
> )
</details>
Added a `RestartAll` method.

---

*[acceptance/terrafarm/farmer.go, line 302 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiNYe4_R1naaTnisCo:-KKizfc1B2etASYVdlab:-591872142) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/terrafarm/farmer.go#L302)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Why not use `Kill()`? You could change the interface so that it takes a boolean `graceful`. If you want to save the boilerplate for iterating over all nodes, you could have a helper `Killall(graceful bool)`.
</details>
I've switched to using `Kill()`. The main concern was that I wanted the `supervisor start cockroach` to happen after `cockroach` has fully stopped, which I believe is guaranteed with `pkill -9` as `Kill()` does. 

---

*[acceptance/terrafarm/farmer.go, line 312 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiOGAFn7vozQUCpnMc:-KKj1lY55EeQ_4m18G0t:610327451) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/terrafarm/farmer.go#L312)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Ditto, this is very close to `ExecRoot` so `ExecRootAll` or remove.
</details>
Removed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
While you're at it, do you think you could make the terraform output less spammy? I think passing `-qy` to the apt-gets will probably get us most of the way. I dread every time I've had to sift through that mess.

---

Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[acceptance/allocator_test.go, line 207 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiKxTccCQvPs2Lw2SH:-KKjIMF92lJma4k5FWO5:-5514012) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L207)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Up-replications (1->3 nodes) are replica adds. Rebalances are adds followed by removes. I *think* splits are relevant, because they tend to trigger adds, but maybe I don't need to concern myself with them, since a lack of replica adds following a split will show up in a higher standard deviation of range counts.
</details>
What I meant was that it would make this test more future proof if you explicitly searched for the event types you're interested in (even though that's currently all of them).

---

*[acceptance/allocator_test.go, line 209 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiKoyANrZ1Jxt0ZNpp:-KKjIZQGR7K8fg-jTrAE:-1190743422) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L209)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> I could've sworn this gave me an error earlier, but it works now. Done.
</details>
Done above, but not here, it seems.

---

*[acceptance/allocator_test.go, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiL9dY5yWH_rf6xibw:-KKjIaINNbHnPDylYmfw:-483822176) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L222)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> I'm returning `stable` below though
</details>
Ah, sorry. Nevermind then.

---

*[acceptance/allocator_terraform/nodectl, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiN1T-ITtg11ADRS_v:-KKjIoQd_Z4L8yhCPQzb:-119573784) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_terraform/nodectl#L1)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Done.
</details>
Remove the whitespace between the shebang and the interpreter.

---

*[acceptance/terrafarm/farmer.go, line 277 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKjJEFBpKQFhnroPKSo:-KKjJEFBpKQFhnroPKSp:-1430255434) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdc672893c720a4df732a42e94ddb07e6bc39a2a/acceptance/terrafarm/farmer.go#L277)):*
> ```Go
> 
> // RestartAll kills then restarts cockroach processes on all nodes.
> func (f *Farmer) RestartAll() error {
> ```

Not to be a stickler, but now there's only one callsite to `RestartAll`, so at this point it seems that you'd want to remove this one as well.
Unrelated: BTW, there seems to be a quickwin in parallelizing the download (and it shouldn't be too much work).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-:-KKjJbAeNjDF6hw5L7nl:1419708597)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[acceptance/allocator_test.go, line 33 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKjJy6JfVqEKFUuKvfS:-KKjJy6JfVqEKFUuKvfT:1770625550) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdc672893c720a4df732a42e94ddb07e6bc39a2a/acceptance/allocator_test.go#L33)):*
> ```Go
> //   -cwd test flag.
> // - You *must* set the TESTTIMEOUT high enough for any of these tests to
> //   finish. To be safe, specify a timeout of at least an hour.
> ```

/an hour/a day/

Didn't you want to mention setting `GOOGLE_PROJECT` as well (among, possibly, other things like which permissions to give to which account)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Would it make sense to print the actual last RangeEvent instead of just these messages:

```
I160620 15:38:36.045002 acceptance/allocator_test.go:227  last range event was 2.636156s ago: waiting for idle allocator
I160620 15:38:46.070485 acceptance/allocator_test.go:227  last range event was 9.171677s ago: waiting for idle allocator
I160620 15:38:56.096704 acceptance/allocator_test.go:227  last range event was 4.911232s ago: waiting for idle allocator
I160620 15:39:06.122227 acceptance/allocator_test.go:227  last range event was 14.93701s ago: waiting for idle allocator
I160620 15:39:16.147604 acceptance/allocator_test.go:227  last range event was 46.976ms ago: waiting for idle allocator
I160620 15:39:26.171381 acceptance/allocator_test.go:227  last range event was 485.497ms ago: waiting for idle allocator
I160620 15:39:36.197796 acceptance/allocator_test.go:227  last range event was 517.269ms ago: waiting for idle allocator
I160620 15:39:46.221917 acceptance/allocator_test.go:227  last range event was 537.371ms ago: waiting for idle allocator
I160620 15:39:56.248182 acceptance/allocator_test.go:227  last range event was 5.542545s ago: waiting for idle allocator
I160620 15:40:06.271529 acceptance/allocator_test.go:227  last range event was 15.564919s ago: waiting for idle allocator
I160620 15:40:16.298537 acceptance/allocator_test.go:227  last range event was 25.591278s ago: waiting for idle allocator
I160620 15:40:26.323721 acceptance/allocator_test.go:227  last range event was 35.616052s ago: waiting for idle allocator
I160620 15:40:36.349894 acceptance/allocator_test.go:227  last range event was 45.641417s ago: waiting for idle allocator
I160620 15:40:46.376177 acceptance/allocator_test.go:227  last range event was 55.667267s ago: waiting for idle allocator
```

cuongdo commented:
I've silenced curl (in `download_binary.sh`) and `apt-get`. Terraform doesn't have similar quiet flags. 

I'm planning on another PR to tackle various improvements, and I'll enhance the log messages as part of that.

---

Review status: 8 of 14 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[acceptance/allocator_test.go, line 207 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiKxTccCQvPs2Lw2SH:-KKjQh-CujAo6j3Q1tzE:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L207)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What I meant was that it would make this test more future proof if you explicitly searched for the event types you're interested in (even though that's currently all of them).
</details>
Done.

---

*[acceptance/allocator_test.go, line 209 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiKoyANrZ1Jxt0ZNpp:-KKjQxLiAgQCkwpBJx9D:971916062) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_test.go#L209)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Done above, but not here, it seems.
</details>
Actually, it looks like I need to revert. This is an error I now get when calling `Scan()` with a `*time.Duration`:

```
E160620 17:04:33.305430 acceptance/allocator_test.go:236  error checking rebalancer: sql: Scan error on column index 0: converting driver.Value type []uint8 ("2.106085s") to a int64: invalid syntax
```

---

*[acceptance/allocator_test.go, line 33 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKjJy6JfVqEKFUuKvfS:-KKjPFaNlOVNoKH9BzT-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdc672893c720a4df732a42e94ddb07e6bc39a2a/acceptance/allocator_test.go#L33)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> /an hour/a day/
> 
> Didn't you want to mention setting `GOOGLE_PROJECT` as well (among, possibly, other things like which permissions to give to which account)?
</details>
Done.

---

*[acceptance/allocator_terraform/nodectl, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKiN1T-ITtg11ADRS_v:-KKjR1FI4URHxZ2IprjN:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a8e844503b918012e412e8d2b90e6b79c110418/acceptance/allocator_terraform/nodectl#L1)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Remove the whitespace between the shebang and the interpreter.
</details>
Done.

---

*[acceptance/terrafarm/farmer.go, line 277 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKjJEFBpKQFhnroPKSo:-KKjRrPThnKde1Z6T8wX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdc672893c720a4df732a42e94ddb07e6bc39a2a/acceptance/terrafarm/farmer.go#L277)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Not to be a stickler, but now there's only one callsite to `RestartAll`, so at this point it seems that you'd want to remove this one as well.
> Unrelated: BTW, there seems to be a quickwin in parallelizing the download (and it shouldn't be too much work).
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-:-KKjT-cDL8GmPlWpkB3v:-1673343233)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
still LGTM. Don't let my latest round of comments trick you into another CI cycle - I assume you'll have a follow-up anyway.

---

Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[acceptance/allocator_test.go, line 107 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKjqiiAOAS-fCyK62GH:-KKjqiiAOAS-fCyK62GI:-1635799847) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef36e66441c008a2d339ef6cf4798b6a4e3bd7a/acceptance/allocator_test.go#L107)):*
> ```Go
> 	defer func() {
> 		if r := recover(); r != nil {
> 			t.Errorf("recovered from panic to destroy cluster")
> ```

should log `r` here, though.

---

*[acceptance/allocator_test.go, line 133 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-KKjqqjyslpBdlFwbr_1:-KKjqqjyslpBdlFwbr_2:-1322072397) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ef36e66441c008a2d339ef6cf4798b6a4e3bd7a/acceptance/allocator_test.go#L133)):*
> ```Go
> 
> 	log.Info("downloading archived stores from Google Cloud Storage in parallel")
> 	var wg sync.WaitGroup
> ```

nit: you don't need a wait group. You can simply feed all returned errors (even when nil) into the channel and loop at the receiver over `0 <= i < NumNodes`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7315#-:-KKjrFbDSwyXK2ewpCgX:-872811152)*
<!-- Sent from Reviewable.io -->
